python/src/nanoarrow/_lib.pyx

```diff
@@ -166,6 +170,70 @@ cdef object alloc_c_array_view(ArrowArrayView** c_array_view):
     return PyCapsule_New(c_array_view[0], 'nanoarrow_array_view', &pycapsule_array_view_deleter)
 
 
+cdef void pycapsule_dlpack_deleter(object dltensor) noexcept:
+    cdef DLManagedTensor* dlm_tensor
+
+    # Do nothing if the capsule has been consumed
+    if PyCapsule_IsValid(dltensor, "used_dltensor"):
+        return
+
+    dlm_tensor = <DLManagedTensor*>PyCapsule_GetPointer(dltensor, 'dltensor')
+    if dlm_tensor == NULL:
+        PyErr_WriteUnraisable(dltensor)
+    # The deleter can be NULL if there is no way for the caller
+    # to provide a reasonable destructor
+    elif dlm_tensor.deleter:
+        dlm_tensor.deleter(dlm_tensor)
+
+
+cdef void view_dlpack_deleter(DLManagedTensor* tensor) noexcept with gil:
+    if tensor.manager_ctx is NULL:
+        return
+    PyMem_Free(tensor)
+    Py_DECREF(<CBufferView>tensor.manager_ctx)
+
+
+cpdef object view_to_dlpack(CBufferView view):
+    cdef DLManagedTensor* dlm_tensor = <DLManagedTensor*>PyMem_Malloc(sizeof(DLManagedTensor))
+
+    cdef DLTensor* dl_tensor = &dlm_tensor.dl_tensor
+    dl_tensor.data = <void*>view._ptr.data.data
+    dl_tensor.ndim = 1
+
+    cdef int64_t* _shape = <int64_t*>PyMem_Malloc(sizeof(int64_t))
+    _shape[0] = len(view)
+    dl_tensor.shape = _shape
+    dl_tensor.strides = NULL
+    dl_tensor.byte_offset = 0
+
+    cdef DLDevice* device = &dl_tensor.device
+    if view._device.device_type_id == ARROW_DEVICE_CPU:
+        device.device_type = kDLCPU
+    else:
+        raise ValueError('Only CPU device is currently supported.')
+    device.device_id =  view._device.device_id
+
+    cdef DLDataType* dtype = &dl_tensor.dtype
+    if view.data_type in ('uint8', 'uint16', 'uint32', 'uint64'):
+        dtype.code = kDLUInt
+    elif view.data_type in ('int8', 'int16', 'int32', 'int64'):
+        dtype.code = kDLInt
+    elif view.data_type in ('half_float', 'float', 'double'):
+        dtype.code = kDLFloat
+    elif view.data_type == 'bool':
+        raise ValueError('Bit-packed boolean data type not supported by DLPack.')
+    else:
+        raise ValueError('DataType is not compatible with DLPack spec: ' + view.data_type)
+    dtype.lanes = <uint16_t>1
+    dtype.bits = <uint8_t>(view.item_size * 8)
+
+    dlm_tensor.manager_ctx = <void*>view
+    Py_INCREF(view)
+    dlm_tensor.deleter = view_dlpack_deleter
```

Comment:
That alone won't help in this case because there is no mechanism right now that this deleter gets called. We would need something similar like `alloc_c_array` that allocates the struct and directly puts it in the capsule with the capsule deleter (which then will call the dlm_tensor deleter)

Reply:
I plan to move most of the code out into a separate method (similarly as how we did it in [Arrow](https://github.com/AlenkaF/arrow/blob/49a978fcef620e6c2f9d68cf3b9ebae03c5a1020/cpp/src/arrow/c/dlpack.cc)). The code will be cleaner and also the checks for the type and device support will be done first and only after that the struct would be allocated.